src/core/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -1,64 +1,105 @@
 /**
- * DISCLAIMER: THIS MIGHT BREAK AT ANY POINT AND/OR NOT WORK FOR CERTAIN PURPOSES
- *
  * This DMA ADC is different than f3xx DMA ADC!
- * This DMA ADC does NOT use the interrupts, as when interrupts were enabled it
- * would constantly interrupt, not letting the print statements or anything
- * else happen.
- * Tried to fix this by changing the sampletime and clock prescaler values, but
- * nothing worked.
- * This was "fixed" by commenting out the NVIC_EnableIRQ, stopping the interrupt
- * from ever being enabled.
+ * f4xx DMA ADC uses a timer to trigger ADC conversions. Timer frequency is 1kHz.
+ *
+ * Timers were added to slow down ADC DMA interrupts, as when they ADC DMA is allowed to convert constantly,
+ * the interrupts stopped the program from doing anything else besides interrupt calls.
  *
- * For commit from before code was removed, refer to commit 81624521a8b2c4b66480193e88cf32782aaee84d.
+ * WARNING: DOES NOT WORK ON PINS THAT ARE ALREADY IN USE (LIKE UART PINS)
  */
 
 #include <HALf4/stm32f4xx.h>
 #include <HALf4/stm32f4xx_hal_adc.h>
 #include <core/io/pin.hpp>
 #include <core/io/platform/f4xx/ADCf4xx.hpp>
 #include <core/io/platform/f4xx/GPIOf4xx.hpp>
+#include <core/platform/f4xx/stm32f4xx.hpp>
+#include <core/utils/log.hpp>
+
+namespace {
+/// This is made as a global variable so that it is accessible in the interrupt.
+DMA_HandleTypeDef* dmaHandle[3];
+ADC_HandleTypeDef* adcHandle[3];
+} // namespace
+
+/**
+ * @brief This function handles DMA2 stream0 global interrupt. (For ADC 1)
+ */
+extern "C" void DMA2_Stream0_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[0]);
+    HAL_ADC_IRQHandler(adcHandle[0]);
+}
+
+/**
+ * @brief This function handles DMA2 stream2 global interrupt. (For ADC 2)
+ */
+extern "C" void DMA2_Stream2_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[1]);
+    HAL_ADC_IRQHandler(adcHandle[1]);
+}
+
+/**
+ * @brief This function handles DMA2 stream1 global interrupt. (For ADC 3)
+ */
+extern "C" void DMA2_Stream1_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[2]);
+    HAL_ADC_IRQHandler(adcHandle[2]);
+}
 
 namespace core::io {
+#define ADC1_SLOT 0
+#define ADC2_SLOT 1
+#define ADC3_SLOT 2
+
+#define ADC1SHIFT 5
+#define ADC2SHIFT 6
+#define ADC3SHIFT 7
 
-// Init static member variables
-ADC_HandleTypeDef ADCf4xx::halADC = {0};
-Pin ADCf4xx::channels[MAX_CHANNELS];
-uint16_t ADCf4xx::buffer[MAX_CHANNELS];
-DMA_HandleTypeDef ADCf4xx::halDMA = {0};
+// Combines the channel memory value with the ADC peripherals it supports into one uint32_t
+#define CHANNEL_SET(adc1, adc2, adc3, ch) (ch | (adc1 << ADC1SHIFT) | (adc2 << ADC2SHIFT) | (adc3 << ADC3SHIFT))
 
-ADCf4xx::ADCf4xx(Pin pin) : ADC(pin) {
-    // Flag representing if the ADC has been configured yet
-    static bool halADCisInit = false;
+ADCf4xx::ADC_State_t ADCf4xx::adcArray[3];
+bool ADCf4xx::timerInit = false;
```

Comment:
so I can't set it to be false in the header because it says "Non-const static data member must be initialized out of line", so i did it there... where should it go

Reply:
I guess this is fine then. I looked it up, and the rules make sense to me